pkg/resource/deploy/plan_executor.go

```diff
@@ -196,3 +208,98 @@ func (pe *planExecutor) handleSingleEvent(event SourceEvent) error {
+	}
+
+	stepExec.SignalCompletion()
+	stepExec.WaitForCompletion()
+
+	if stepExec.Errored() {
+		return execError("failed", preview)
+	} else if canceled {
+		return execError("canceled", preview)
+	}
+
+	// Rebuild this plan's map of old resources and dependency graph, stripping out any deleted resources and repairing
+	// dependency lists as necessary. Note that this updates the base snapshot _in memory_, so it is critical that any
+	// components that use the snapshot refer to the same instance and avoid reading it concurrently with this rebuild.
+	//
+	// The process of repairing dependency lists is a bit subtle. Because multiple physical resources may share a URN,
+	// the ability of a particular URN to be referenced in a dependency list can change based on the dependent
+	// resource's position in the resource list. For example, consider the following list of resources, where each
+	// resource is a (URN, ID, Dependencies) tuple:
+	//
+	//     [ (A, 0, []), (B, 0, [A]), (A, 1, []), (A, 2, []), (C, 0, [A]) ]
+	//
+	// Let `(A, 0, [])` and `(A, 2, [])` be deleted by the refresh. This produces the following intermediate list
+	// before dependency lists are repaired:
+	//
+	//     [ (B, 0, [A]), (A, 1, []), (C, 0, [A]) ]
+	//
+	// In order to repair the dependency lists, we iterate over the intermediate resource list, keeping track of which
+	// URNs refer to at least one physical resource at each point in the list, and remove any dependencies that refer
+	// to URNs that do not refer to any physical resources. This process produces the following final list:
+	//
+	//     [ (B, 0, []), (A, 1, []), (C, 0, [A]) ]
+	//
+	// Note that the correctness of this process depends on the fact that the list of resources is a topological sort
+	// of its corresponding dependency graph, so a resource always appears in the list after any resources on which it
+	// may depend.
+	resources := make([]*resource.State, 0, len(prev.Resources))
+	referenceable := make(map[resource.URN]struct{})
+	olds := make(map[resource.URN]*resource.State)
+	for _, s := range steps {
+		new := s.New()
+		if new == nil {
+			contract.Assert(s.Old().Custom)
+			contract.Assert(!providers.IsProviderType(s.Old().Type))
+			continue
+		}
+
+		// Remove any deleted resources from this resource's dependency list.
+		if len(new.Dependencies) != 0 {
+			deps := make([]resource.URN, 0, len(new.Dependencies))
+			for _, d := range new.Dependencies {
+				if _, canRef := referenceable[d]; canRef {
+					deps = append(deps, d)
+				}
+			}
+			new.Dependencies = deps
+		}
+
+		// Add this resource to the resource list and mark it as referenceable.
+		resources = append(resources, new)
+		referenceable[new.URN] = struct{}{}
+
+		// Do not record resources that are pending deletion in the "olds" lookup table.
+		if !new.Delete {
+			olds[new.URN] = new
+		}
+	}
+	pe.plan.prev.Resources = resources
+	pe.plan.olds, pe.plan.depGraph = olds, graph.NewDependencyGraph(resources)
```

Comment:
This is pretty subtle, do you mind adding a note in `deploy.NewPlan` that these things can get mutated in the case of refreshes?

Reply:
Sure. I also considered moving that logic wholesale into the plan executor, but it's not uniform (i.e. it's different between cases where we do and do not perform a refresh). That said, do you think that would be clearer?

Reply:
Hmm, I think it's clearer here than it would be anywhere else.

Reply:
To be clear, I thought about moving the logic in `deploy.NewPlan` that computes these things into `planExecutor.Execute`.

Reply:
I'm happy to see all of the `if refresh {` go away in this PR and I think it would be a shame to re-introduce it in other places when we can do what we're doing here and keep the code separated.

Reply:
> To be clear, I thought about moving the logic in deploy.NewPlan that computes these things into planExecutor.Execute.

I think I'm fine either way - up to you.